### PR TITLE
Add pstools config file.

### PR DESCRIPTION
**Pull request by junaruga**
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [X] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

---

CC: @shilpagarg

This PR is to add the bio tool called [pstools]https://github.com/shilpagarg/pstools) to Galaxy tool. Actually the actual source code is not merged to the main repository. We used the source code on my forked repository https://github.com/junaruga/pstools/tree/wip/staging .

The software is currently available on the container registry https://github.com/users/junaruga/packages/container/package/pstools . Eventually we want to enable the software as bioconda package and BioContainers container repository. But as our first step to enable it to the tool. We want to use the GitHub container repository.

Showing the help.

```
$ docker run --rm -t ghcr.io/junaruga/pstools:latest
Usage: pstools <command> <arguments> <inputs>
Commands:
  clean_graph           cleaning the given graph
  intersect             extract corresponding paf records for sub graph
  resolve_haplotypes    use hic data to resolve haplotypes
  haplotype_scaffold    scaffolding the predicted haplotypes
  hic_mapping_unitig    map hic data to sequences in the graph
  hic_mapping_haplo     map hic data to predicted haplotypes for scaffolding
  qv                    calculate qv score for prediction
  completeness          calculate completeness for prediction
  phasing_error         calculate switch error
  version               print version number
```

Run the following command to run qv score to get the output file.

```
$ docker run --rm -t ghcr.io/junaruga/pstools:latest qv seq.fa hic.R1.fastq.gz hic.R2.fastq.gz > qv.txt

$ cat qv.txt
```

## Notes

* We should prepare the test data ideally. But right now I don't have it.

This is my first PR for the Galaxy project. Any feedback is welcome. Thank you.

### CODE REVIEW

File: tools/pstools/pstools.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<tool id="pstools_qv" name="Calculate qv score for prediction (pstools)" version="0.1.0" python_template_version="3.5">
```
Comment (junaruga) [inline tools/pstools/pstools.xml:1]: Here the version is set as `0.1.0`, but the actually used source is on the branch https://github.com/junaruga/pstools/tree/wip/staging . Maybe we need to merge the branch to the original repo, and need to do `git tag` before merging this PR?

File: tools/pstools/pstools.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<tool id="pstools_qv" name="Calculate qv score for prediction (pstools)" version="0.2a1" python_template_version="3.5">
```
Comment (junaruga) [inline tools/pstools/pstools.xml:1]: The used source code version is `v0.2a1`.
https://github.com/junaruga/pstools/tree/v0.2a1

**Comment by junaruga** [comment]
I am not sure if using container to get the package is allowed to be merged to this repository. So, just in case, I am preparing the bioconda package here: https://github.com/bioconda/bioconda-recipes/pull/31472 .

File: tools/pstools/pstools.xml (hunk 2 above)
Comment (bgruening) [reply tools/pstools/pstools.xml:1]: python_template_version is not really needed, instead please add a "profile=20.05"

File: tools/pstools/pstools.xml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    <requirements>
+        <container type="docker">ghcr.io/junaruga/pstools:latest</container>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+        pstools qv '$input1' '$input2' '$input3' > '$output1'
+    ]]></command>
+    <inputs>
+        <param type="data" name="input1" format="fa" />
```
Comment (bgruening) [inline tools/pstools/pstools.xml:9]: Those need to be proper Galaxy datatypes. You can see them here: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/datatypes_conf.xml.sample

Comment (bgruening) [reply tools/pstools/pstools.xml:9]: please also add a label and ideally also a help

Comment (bgruening) [reply tools/pstools/pstools.xml:9]: I would also change the names, to something meaningful.

File: tools/pstools/pstools.xml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        <param type="data" name="input2" format="gz" />
+        <param type="data" name="input3" format="gz" />
+    </inputs>
+    <outputs>
+        <data name="output1" format="txt" />
+    </outputs>
+    <help><![CDATA[
+        pstools qv --help
```
Comment (bgruening) [inline tools/pstools/pstools.xml:17]: please add a proper help that is meaningful for biologists :)

**Review by bgruening — COMMENTED** [review]
Hi @junaruga! Thanks for your contribution. A few comments inline and a few general comments:

* tests are missing, please use `planemo test` and `planemo serve`
* help is missing
* a conda package is indeed better :)

File: tools/pstools/pstools.xml (hunk 3 above)
Comment (junaruga) [reply tools/pstools/pstools.xml]: I updated as a proper datatype, I added the label, and I added the proper help content. I change the names to the meaningful ones.

File: tools/pstools/pstools.xml (hunk 4 above)
Comment (junaruga) [reply tools/pstools/pstools.xml]: I added the meaningful help content.

**Comment by junaruga** [comment]
I rebased.

## planemo lint

I see some warnings.

```
$ pwd
/home/jaruga/git/galaxyproject/tools-iuc/tools/pstools

$ planemo lint
Linting tool /home/jaruga/var/git/galaxyproject/tools-iuc/tools/pstools/pstools.xml
Applying linter tests... CHECK
.. CHECK: 1 test(s) found.
Applying linter output... CHECK
.. INFO: 1 outputs found.
Applying linter inputs... CHECK
.. INFO: Found 3 input parameters.
Applying linter help... WARNING
.. WARNING: Invalid reStructuredText found in help - [<string>:10: (ERROR/3) Unexpected indentation.
].
.. CHECK: Tool contains help section.
Applying linter general... WARNING
.. WARNING: Tool version [@TOOL_VERSION@+galaxy@VERSION_SUFFIX@] is not compliant with PEP 440.
.. CHECK: Tool defines a name [Calculate qv score for prediction (pstools)].
.. CHECK: Tool defines an id [pstools].
.. CHECK: Tool targets 16.01 Galaxy profile.
Applying linter command... CHECK
.. INFO: Tool contains a command.
Applying linter citations... CHECK
.. CHECK: Found 1 likely valid citations.
Applying linter tool_xsd... CHECK
.. INFO: File validates against XML schema.
Failed linting
```

For the Tool version, I referred the following code to use same version of the pstools for this XML file's version.
https://github.com/galaxyproject/tools-iuc/blob/febca656bf539c044816abebfeef3c84adc4056e/tools/minimap2/minimap2.xml#L2

Do you know how to fix the warnings?

## planemo test


The test is success.
But there is a difference between the output files in my local (Fedora Linux 34) and in Galaxy executed by `planemo test`. Right now I adapted the content of output in the Galaxy to pass the test of the content of the output

```
--- local_file
+++ history_data
@@ -1,2 +1,2 @@
 QV_RAW>QV
--1.000>-1.000
+19.628>-1.000
```

## planemo serve

Success!

**Comment by junaruga** [comment]
> But there is a difference between the output files in my local (Fedora Linux 34) and in Galaxy executed by planemo test. Right now I adapted the content of output in the Galaxy to pass the test of the content of the output

There was a mistake about how to run the `pstools` on my local. Just one of the command arguments were wrong. After correcting it, I see the same result from the Galaxy.



```
$ ../../pstools qv ./true_hap2.fasta ./hap2.1.fastq ./hap2.2.fastq
[M::worker_pipeline::0.180*1.67] processed 10 sequences; 727 distinct k-mers in the hash table
[M::worker_pipeline::0.277*1.67] processed 10 sequences; 727 distinct k-mers in the hash table
[M::yak_qv_cb] read 1 sequences
[M::yak_qv_cb] read 0 sequences
[M::yak_qv_cb@0.31*1.88] processed 1 sequences
[M::yak_qv_cb] read 0 sequences

[M::Result] QV_RAW  QV
19.628  -1.000
```

When running the bioconda `pstools` on my Fedora 34, I see the Segmentation fault, but it is another issue.

```
(base) $ which pstools
~/miniconda3/bin/pstools
(base) $ pstools qv ./true_hap2.fasta.rb ./hap2.1.fastq ./hap2.2.fastq
[M::worker_pipeline::0.172*1.59] processed 10 sequences; 727 distinct k-mers in the hash table
[M::worker_pipeline::0.259*1.61] processed 10 sequences; 727 distinct k-mers in the hash table
Segmentation fault (core dumped)
```

**Comment by junaruga** [comment]
The `planemo lint` failed on CI. I need to fix it.
https://github.com/galaxyproject/tools-iuc/runs/4168541439?check_suite_focus=true

**Comment by junaruga** [comment]
It seems space indented lines after `Commands:` and `Options:` are something wrong. When removed the lines, the lint passed.

**Comment by junaruga** [comment]
Now I added one more commit to fix the `planemo lint`. Here is the difference.

```
diff --git a/tools/pstools/pstools.xml b/tools/pstools/pstools.xml
index d7f559578..432458d03 100644
--- a/tools/pstools/pstools.xml
+++ b/tools/pstools/pstools.xml
@@ -1,4 +1,4 @@
-<tool id="pstools" name="Calculate qv score for prediction (pstools)" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" python_template_version="3.5">
+<tool id="pstools" name="Calculate qv score for prediction (pstools)" version="0.2a3" python_template_version="3.5">
     <requirements>
         <requirement type="package" version="0.2a3">pstools</requirement>
     </requirements>
@@ -41,22 +41,23 @@ In this tool, the `pstols gv` command is avalable.
 
 Usage: pstools <command> <arguments> <inputs>
 Commands:
-  clean_graph           cleaning the given graph
-  intersect             extract corresponding paf records for sub graph
-  resolve_haplotypes    use hic data to resolve haplotypes
-  haplotype_scaffold    scaffolding the predicted haplotypes
-  hic_mapping_unitig    map hic data to sequences in the graph
-  hic_mapping_haplo     map hic data to predicted haplotypes for scaffolding
-  qv                    calculate qv score for prediction
-  completeness          calculate completeness for prediction
-  phasing_error         calculate switch error
-  version               print version number
+* clean_graph           cleaning the given graph
+
+* intersect             extract corresponding paf records for sub graph
+* resolve_haplotypes    use hic data to resolve haplotypes
+* haplotype_scaffold    scaffolding the predicted haplotypes
+* hic_mapping_unitig    map hic data to sequences in the graph
+* hic_mapping_haplo     map hic data to predicted haplotypes for scaffolding
+* qv                    calculate qv score for prediction
+* completeness          calculate completeness for prediction
+* phasing_error         calculate switch error
+* version               print version number
 
 Usage: pstools qv [options] <seq.fa> <hic.R1.fastq.gz> <hic.R2.fastq.gz>
 Options:
-  -k INT     k-mer size [31]
-  -t INT     number of worker threads [4]
-  -K INT     chunk size [100m]
+* -k INT     k-mer size [31]
+* -t INT     number of worker threads [4]
+* -K INT     chunk size [100m]
     ]]></help>
     <citations>
         <citation type="bibtex">
```

**Comment by junaruga** [comment]
> python_template_version is not really needed, instead please add a "profile=20.05"

I missed this comment. I will fix it.

**Comment by junaruga** [comment]
> I missed this comment. I will fix it.

Rebased and squashed with 1 commit now.

File: tools/pstools/pstools.xml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      <regex match="Exception"/>
+      <regex match="Segfault"/>
+    </stdio>
+    <command><![CDATA[
+        pstools qv '$fasta_input1' '$fastq_input1' '$fastq_input2' > '$output1' || :
+    ]]></command>
+
+    <inputs>
+        <param type="data" name="fasta_input1" label="Reference fasta file" format="fasta" />
+        <param type="data" name="fastq_input1" label="Read 1 fastq file" format="fastq" />
```
Comment (bgruening) [inline tools/pstools/pstools.xml:21]: If you wanna go fancier, you could support here a paired-end collection as well. Let me know if you want to explore this or if we should get this merged.

File: tools/pstools/pstools.xml [hunk 6]
```diff
@@ -0,0 +1,71 @@
+<tool id="pstools" name="Calculate qv score for prediction (pstools)" version="0.2a3" profile="20.05">
```
Comment (bgruening) [inline tools/pstools/pstools.xml:1]: keep the "name" short, but add a "description"

File: tools/pstools/pstools.xml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+* hic_mapping_haplo     map hic data to predicted haplotypes for scaffolding
+* qv                    calculate qv score for prediction
+* completeness          calculate completeness for prediction
+* phasing_error         calculate switch error
+* version               print version number
+
+Usage: pstools qv [options] <seq.fa> <hic.R1.fastq.gz> <hic.R2.fastq.gz>
+Options:
+* -k INT     k-mer size [31]
+* -t INT     number of worker threads [4]
```
Comment (bgruening) [inline tools/pstools/pstools.xml:59]: this parameter you can add into the command section with GALAXY_SLOTS

File: tools/pstools/pstools.xml [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+* hic_mapping_unitig    map hic data to sequences in the graph
+* hic_mapping_haplo     map hic data to predicted haplotypes for scaffolding
+* qv                    calculate qv score for prediction
+* completeness          calculate completeness for prediction
+* phasing_error         calculate switch error
+* version               print version number
+
+Usage: pstools qv [options] <seq.fa> <hic.R1.fastq.gz> <hic.R2.fastq.gz>
+Options:
+* -k INT     k-mer size [31]
```
Comment (bgruening) [inline tools/pstools/pstools.xml:58]: why was this not added?

Comment (junaruga) [reply tools/pstools/pstools.xml:58]: I didn't consider to add the `pstools qv` options such as `-k`. You want me to add the options too to the galaxy tool, right?

Comment (shilpagarg) [reply tools/pstools/pstools.xml:58]: ``-k`` is the default.

Comment (bgruening) [reply tools/pstools/pstools.xml:58]: I think that would be nice. If not we should not confuse the users with this help text, if you can not set those options.

Comment (shilpagarg) [reply tools/pstools/pstools.xml:58]: Let's wait a bit before you can merge the PR. Thanks for your feedback.

**Comment by bgruening** [comment]
@junaruga can you please also add a .shed.yml file? Thanks.

**Comment by junaruga** [comment]
> can you please also add a .shed.yml file? Thanks.

Thanks for the review! OK. I will do it possibly in a few weeks.

**Comment by junaruga** [comment]
Sorry I won't work for this PR any more. Please open another PR if you are still interested in it.

**Pull request closed without merging**